flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -5366,27 +5448,123 @@ class ArrayExprLowering {
-            [&](const Fortran::evaluate::StaticDataObject::Pointer &) -> CC {
-              fir::emitFatalError(loc, "substring of static array object");
-            }},
-        x.parent());
-    // lower and upper *must* be scalars
-    llvm::SmallVector<mlir::Value> bounds = {fir::getBase(asScalar(x.lower()))};
-    if (auto upper = x.upper())
-      bounds.push_back(fir::getBase(asScalar(*upper)));
-    return [=](IterSpace iters) -> ExtValue {
+    if (isProjectedCopyInCopyOut()) {
+      ccStoreToDest = [=, &x](IterSpace iters) {
+        auto loc = getLoc();
+        auto innerArg = iters.innerArgument();
+        auto exv = iters.elementExv();
+        auto arrTy = innerArg.getType();
+        auto eleTy = fir::applyPathToType(arrTy, iters.iterVec());
+        auto resRefTy = builder.getRefType(eleTy);
+        // Get a reference to the array element to be amended.
+        auto arrayOp = builder.create<fir::ArrayAccessOp>(
+            loc, resRefTy, innerArg, iters.iterVec(), destination.typeparams());
+        auto charTy = eleTy.dyn_cast<fir::CharacterType>();
+        assert(charTy && "must be character");
+        llvm::SmallVector<mlir::Value> bounds = {
+            fir::getBase(asScalar(x.lower()))};
+        if (auto upper = x.upper())
+          bounds.push_back(fir::getBase(asScalar(*upper)));
+        fir::factory::CharacterExprHelper helper{builder, loc};
+        auto len = fir::getLen(exv);
+        auto charBox =
+            helper.createSubstring(fir::CharBoxValue{arrayOp, len}, bounds);
+        // Generate an optimized CHARACTER assignment.
+        fir::factory::genCharacterCopy(charBox.getAddr(), charBox.getLen(),
+                                       fir::getBase(exv), fir::getLen(exv),
+                                       builder, loc);
+        // Mark this array element as amended.
+        auto ty = innerArg.getType();
+        auto amend = builder.create<fir::ArrayAmendOp>(loc, ty, innerArg,
+                                                       charBox.getAddr());
+        return abstractArrayExtValue(amend, len);
+      };
+    }
+    auto pf = std::visit([&](const auto &v) { return genarr(v); }, x.parent());
+    if (!isBoxValue())
+      return pf;
+    return [=, &x](IterSpace iters) -> ExtValue {
+      // Lower and upper expressions must be scalars.
+      llvm::SmallVector<mlir::Value> bounds = {
+          fir::getBase(asScalar(x.lower()))};
+      if (auto upper = x.upper())
+        bounds.push_back(fir::getBase(asScalar(*upper)));
       auto base = pf(iters);
-      if (auto *chr = base.getCharBox())
-        return fir::factory::CharacterExprHelper{builder, loc}.createSubstring(
-            *chr, bounds);
-      TODO(loc, "unhandled substring base type");
-      return mlir::Value{};
+      return base.match(
+          [&](const fir::BoxValue &box) -> fir::ExtendedValue {
+            if (auto embox = mlir::dyn_cast_or_null<fir::EmboxOp>(
```

Comment:
Why not using a rebox here to cover all cases ? (This assumes rebox support for fir.slice with a substring).
Merging consecutive embox and rebox could be a FIR pass.